processor/src/test/java/org/mapstruct/ap/test/bugs/_3591/CompilationErrorDuplicateMethodsBugTest.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._3591;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.runner.GeneratedSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@IssueKey("3591")
+public class CompilationErrorDuplicateMethodsBugTest {
+
+    @RegisterExtension
+    GeneratedSource generatedSource = new GeneratedSource();
+
+    @ProcessorTest
+    @WithClasses({
+        BeanDto.class,
+        Bean.class,
+        BeanMapper.class
+    })
+    void mapNestedBeansWithMappingAnnotation() {
+        Bean bean = testBean();
+
+        BeanDto beanDto = BeanMapper.INSTANCE.map( bean, new BeanDto() );
+
+        assertThat( beanDto.getValue() ).isEqualTo( "parent" );
+        assertThat( beanDto.getBeans() ).isNotEmpty();
+        assertThat( beanDto.getBeans().get( 0 ).getValue() ).isEqualTo( "child" );
+    }
+
+    @ProcessorTest
+    @WithClasses({
+        ContainerBean.class,
+        ContainerBeanDto.class,
+        ContainerBeanMapper.class,
+    })
+    void shouldCreateOnlyOneMethodForStreamMapping() {
+        generatedSource.addComparisonToFixtureFor( ContainerBeanMapper.class );
+    }
+
+    @ProcessorTest
+    @WithClasses({
+        ContainerBean.class,
+        ContainerBeanDto.class,
+        ContainerBeanMapper.class,
+    })
+    void shouldMapNestedMapAndStream() {
+        ContainerBean containerBean = testContainerBean();
+
+        ContainerBeanDto dto = ContainerBeanMapper.INSTANCE.mapWithMapMapping( containerBean, new ContainerBeanDto() );
+
+        assertThat( dto.getBeanMap() ).isNotEmpty();
+        assertThat( dto.getBeanMap().get( "child" ).getValue() ).isEqualTo( "mapChild" );
+
+        List<ContainerBeanDto> fromStream = dto.getBeanStream().collect( Collectors.toList() );
+        assertThat( fromStream ).isNotEmpty();
+        assertThat( fromStream.get( 0 ).getValue() ).isEqualTo( "streamChild" );
+    }
+
+    static ContainerBean testContainerBean() {
```

Comment:
Should these helper methods really be own methods? They are only used 1 time in their respective tests.